### PR TITLE
Add features to azure_rm_appgateway

### PR DESCRIPTION
**Pull request by petitalotc**
##### SUMMARY

- [x] add waf configuration settings
- [x] add trusted root certificates 
- [x] add enable http 2 parameter
- [x] add autoscale configuration
- [x] add draining options

All of these features are provided by the Azure python SDK however was not previously included in the azure_rm_appgateway module.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

azure_rm_appgateway.py

##### ADDITIONAL INFORMATION

integration tests. results 
```
ansible-test integration azure_rm_appgateway --allow-destructive -v --allow-disabled
Using existing azure cloud config: tests/integration/cloud-config-azure.ini
Creating container database.
Configuring target inventory.
Running azure_rm_appgateway integration test role
Initializing "/tmp/ansible-test-scmqo3l5-injector" as the temporary injector directory.
Stream command: ansible-playbook azure_rm_appgateway-ma6xzyni.yml -i inventory -v
[WARNING]: running playbook inside collection azure.azcollection

PLAY [testhost] ****************************************************************

TASK [Gathering Facts] *********************************************************
ok: [testhost]

TASK [azure_rm_appgateway : Prepare random number] *****************************
ok: [testhost] => {"ansible_facts": {"cert1_file": "cert1.txt", "cert2_file": "cert2.txt", "cert3b64_file": "cert3b64.txt", "rpfx": "9f0db4438"}, "changed": false}

TASK [azure_rm_appgateway : Set json query to only retrieve gateways from this test] ***
ok: [testhost] => {"ansible_facts": {"query": "[?ends_with(name, `9f0db4438`)]"}, "changed": false}

TASK [azure_rm_appgateway : Load app gateways] *********************************
ok: [testhost] => {"changed": false, "gateways": [{"id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/applicationGateways/appgateway9f0db44709", "location": "eastus", "name": "appgateway9f0db44709", "operational_state": "Running", "provisioning_state": "Succeeded", "resource_group": "rg-agaetis-ansible-azure-contrib", "ssl_policy": {"policy_name": "ssl_policy20150501", "policy_type": "predefined"}}, {"id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/applicationGateways/appgateway9f0db44566", "location": "eastus", "name": "appgateway9f0db44566", "operational_state": "Running", "provisioning_state": "Succeeded", "resource_group": "rg-agaetis-ansible-azure-contrib", "ssl_policy": {"policy_name": null, "policy_type": "custom"}}, {"id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/applicationGateways/appgateway-waf-v2-9f0db44989", "location": "eastus", "name": "appgateway-waf-v2-9f0db44989", "operational_state": "Running", "provisioning_state": "Succeeded", "resource_group": "rg-agaetis-ansible-azure-contrib", "ssl_policy": {"policy_name": "ssl_policy20170401_s", "policy_type": "predefined"}}, {"id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/applicationGateways/appgateway-v2-9f0db44989", "location": "eastus", "name": "appgateway-v2-9f0db44989", "operational_state": "Running", "provisioning_state": "Succeeded", "resource_group": "rg-agaetis-ansible-azure-contrib", "ssl_policy": {"policy_name": "ssl_policy20170401_s", "policy_type": "predefined"}}, {"id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/applicationGateways/appgateway-subnet-lookup9f0db44566", "location": "eastus", "name": "appgateway-subnet-lookup9f0db44566", "operational_state": "Running", "provisioning_state": "Succeeded", "resource_group": "rg-agaetis-ansible-azure-contrib", "ssl_policy": {"policy_name": "ssl_policy20150501", "policy_type": "predefined"}}]}

TASK [azure_rm_appgateway : Assert there are no gateways] **********************
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [azure_rm_appgateway : Create a virtual network] **************************
changed: [testhost] => {"changed": true, "check_mode": false, "state": {"address_prefixes": ["10.1.0.0/16", "172.100.0.0/16"], "dns_servers": ["127.0.0.1", "127.0.0.2"], "etag": "W/\"715224cb-95af-401c-904f-e2f74199eb1f\"", "id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/virtualNetworks/vnet9f0db4438", "location": "eastus", "name": "vnet9f0db4438", "provisioning_state": "Succeeded", "tags": null, "type": "Microsoft.Network/virtualNetworks"}}

TASK [azure_rm_appgateway : Create a subnet] ***********************************
changed: [testhost] => {"changed": true, "state": {"address_prefix": "10.1.0.0/24", "address_prefixes": null, "id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/virtualNetworks/vnet9f0db4438/subnets/subnet9f0db4438", "name": "subnet9f0db4438", "nat_gateway": null, "network_security_group": {}, "private_endpoint_network_policies": "Enabled", "private_link_service_network_policies": "Enabled", "provisioning_state": "Succeeded", "route_table": {}}}

TASK [azure_rm_appgateway : Configure public IP for v2 gateway] ****************
changed: [testhost] => {"changed": true, "state": {"dns_settings": {}, "etag": "W/\"725019e4-2a09-403c-b0fe-cef489e06104\"", "idle_timeout_in_minutes": 4, "ip_address": "20.168.220.142", "location": "eastus", "name": "appgateway-v2-9f0db4438-pip", "provisioning_state": "Succeeded", "public_ip_address_version": "ipv4", "public_ip_allocation_method": "static", "sku": "Standard", "tags": null, "type": "Microsoft.Network/publicIPAddresses", "zones": null}}

TASK [azure_rm_appgateway : Try to create v2 instance of Application Gateway with autoscale configuration and trusted root certificates] ***
changed: [testhost] => {"changed": true, "id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/applicationGateways/appgateway-v2-9f0db4438", "location": "eastus", "name": "appgateway-v2-9f0db4438", "operational_state": "Running", "provisioning_state": "Succeeded", "resource_group": "rg-agaetis-ansible-azure-contrib"}

TASK [azure_rm_appgateway : Assert the resource instance is well created] ******
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [azure_rm_appgateway : Try to create v2 instance of Application Gateway with autoscale configuration and trusted root certificates - no update] ***
ok: [testhost] => {"changed": false, "id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/applicationGateways/appgateway-v2-9f0db4438", "location": "eastus", "name": "appgateway-v2-9f0db4438", "operational_state": "Running", "provisioning_state": "Succeeded", "resource_group": "rg-agaetis-ansible-azure-contrib"}

TASK [azure_rm_appgateway : Assert the resource instance is not updated] *******
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [azure_rm_appgateway : Configure public IP for waf_v2 gateway] ************
changed: [testhost] => {"changed": true, "state": {"dns_settings": {}, "etag": "W/\"3e142ecb-57e8-48e5-84b6-14c274f3f8bf\"", "idle_timeout_in_minutes": 4, "ip_address": "52.188.163.254", "location": "eastus", "name": "appgateway-waf-v2-9f0db4438-pip", "provisioning_state": "Succeeded", "public_ip_address_version": "ipv4", "public_ip_allocation_method": "static", "sku": "Standard", "tags": null, "type": "Microsoft.Network/publicIPAddresses", "zones": null}}

TASK [azure_rm_appgateway : Try to create waf_v2 instance of Application Gateway with waf configuration] ***
changed: [testhost] => {"changed": true, "id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/applicationGateways/appgateway-waf-v2-9f0db4438", "location": "eastus", "name": "appgateway-waf-v2-9f0db4438", "operational_state": "Running", "provisioning_state": "Succeeded", "resource_group": "rg-agaetis-ansible-azure-contrib"}

TASK [azure_rm_appgateway : Assert the resource instance is well created] ******
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [azure_rm_appgateway : Try to create waf_v2 instance of Application Gateway with waf configuration - no update] ***
ok: [testhost] => {"changed": false, "id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/applicationGateways/appgateway-waf-v2-9f0db4438", "location": "eastus", "name": "appgateway-waf-v2-9f0db4438", "operational_state": "Running", "provisioning_state": "Succeeded", "resource_group": "rg-agaetis-ansible-azure-contrib"}

TASK [azure_rm_appgateway : Assert the resource instance is not updated] *******
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [azure_rm_appgateway : Delete v2 instance of Application Gateway] *********
changed: [testhost] => {"changed": true}

TASK [azure_rm_appgateway : Assert the state has changed] **********************
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [azure_rm_appgateway : Delete waf_v2 instance of Application Gateway] *****
changed: [testhost] => {"changed": true}

TASK [azure_rm_appgateway : Assert the state has changed] **********************
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [azure_rm_appgateway : Delete public IP for v2 gateway] *******************
changed: [testhost] => {"changed": true, "state": {"status": "Deleted"}}

TASK [azure_rm_appgateway : Assert the state has changed] **********************
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [azure_rm_appgateway : Delete public IP for waf v2 gateway] ***************
changed: [testhost] => {"changed": true, "state": {"status": "Deleted"}}

TASK [azure_rm_appgateway : Assert the state has changed] **********************
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [azure_rm_appgateway : Delete virtual network] ****************************
changed: [testhost] => {"changed": true, "check_mode": false, "state": {"address_prefixes": ["10.1.0.0/16", "172.100.0.0/16"], "dns_servers": ["127.0.0.1", "127.0.0.2"], "etag": "W/\"542ab011-19b1-49b5-b54c-a38f1cf45b3a\"", "id": "/subscriptions/4aea99db-397a-45e5-a508-e6e06db77164/resourceGroups/rg-agaetis-ansible-azure-contrib/providers/Microsoft.Network/virtualNetworks/vnet9f0db4438", "location": "eastus", "name": "vnet9f0db4438", "provisioning_state": "Succeeded", "status": "Deleted", "tags": null, "type": "Microsoft.Network/virtualNetworks"}}

PLAY RECAP *********************************************************************
testhost                   : ok=26   changed=11   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

### CODE REVIEW

**Comment by petitalotc** [comment]
It also fixed this [issues/478](https://github.com/ansible-collections/azure/issues/478)

File: plugins/modules/azure_rm_appgateway.py [hunk 1]
```diff
@@ -2022,10 +2364,40 @@ def rerite_rule_set_id(subscription_id, resource_group_name, appgateway_name, na
...
+    """Compare two dictionaries using recursive_diff method and assuming that null values coming from yaml input
+    are acting like absent values"""
+    oldd = old_params.get(param_name, {})
+    newd = new_params.get(param_name, {})
+
+    if oldd == {} and newd == {}:
+        return True
+
+    diffs = recursive_diff(oldd, newd)
+    if diffs == None:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]: Comparison 'diffs == None' should be 'diffs is None'

File: plugins/modules/azure_rm_appgateway.py [hunk 2]
```diff
@@ -2022,10 +2364,40 @@ def rerite_rule_set_id(subscription_id, resource_group_name, appgateway_name, na
...
+        return True
+    else:
+        actual_diffs = diffs[1]
+        return all(value == None or not value for value in actual_diffs.values())
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]: Comparison 'diffs == None' should be 'diffs is None'

File: plugins/modules/azure_rm_appgateway.py [hunk 3]
```diff
@@ -612,6 +652,101 @@
...
+                        description:
+                            - The variable to be excluded.
+                        type: str
+                    selector:
+                        description:
+                            - When match_variable is a collection, operator used to specify which elements in the collection this exclusion applies to.
+                        type: str
+                    selector_match_operator:
+                        description:
+                            - When match_variable is a collection, operate on the selector to specify which elements in the collection this exclusion applies to.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]: line too long (161 > 160 characters)

File: plugins/modules/azure_rm_appgateway.py [hunk 4]
```diff
@@ -676,6 +814,53 @@
...
+    frontend_ip_configurations:
+      - subnet:
+          id: "{{ subnet_id }}"
+        name: sample_gateway_frontend_ip_config
+    frontend_ports:
+      - port: 90
+        name: ag_frontend_port
+    trusted_root_certificates:
+      - name: "root_cert"
+        key_vault_secret_id: "https://kv/secret"
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]:
```suggestion
        key_vault_secret_id: "https://kv/secret"
```

File: plugins/modules/azure_rm_appgateway.py [hunk 5]
```diff
@@ -1037,6 +1222,94 @@
         include_query_string: true
         url_path_maps:
           - "path_mappings"
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]: blank line contains whitespace

```suggestion

```

File: plugins/modules/azure_rm_appgateway.py [hunk 6]
```diff
@@ -1037,6 +1222,94 @@
...
+      - frontend_ip_configuration: sample_gateway_frontend_ip_config
+        frontend_port: ag_frontend_port
+        name: sample_http_listener
+    request_routing_rules:
+      - rule_type: Basic
+        backend_address_pool: test_backend_address_pool
+        backend_http_settings: sample_appgateway_http_settings
+        http_listener: sample_http_listener
+        name: rule1
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]: blank line contains whitespace

```suggestion

```

File: plugins/modules/azure_rm_appgateway.py [hunk 7]
```diff
@@ -1037,6 +1222,94 @@
...
+        frontend_port: ag_frontend_port
+        name: sample_http_listener
+    request_routing_rules:
+      - rule_type: Basic
+        backend_address_pool: test_backend_address_pool
+        backend_http_settings: sample_appgateway_http_settings
+        http_listener: sample_http_listener
+        name: rule1
+        
+- name: Create instance of Application Gateway waf_v2 with waf configuration
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]:
```suggestion
- name: Create instance of Application Gateway waf_v2 with waf configuration
```

File: plugins/modules/azure_rm_appgateway.py [hunk 8]
```diff
@@ -1518,6 +1844,14 @@ def exec_module(self, **kwargs):
                                           kwargs['name'],
                                           item['probe'])
                             item['probe'] = {'id': id}
+                        if 'trusted_root_certificates' in item:
+                            for j in range(len(item['trusted_root_certificates'])):
+                                id = item['trusted_root_certificates'][j]
+                                id = id if is_valid_resource_id(id) else trusted_root_certificate_id(self.subscription_id,
+                                                                                             kwargs['resource_group'],
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]: line 1851 and 1852 continuation line under-indented for visual indent

File: plugins/modules/azure_rm_appgateway.py [hunk 9]
```diff
@@ -1596,7 +1930,7 @@ def exec_module(self, **kwargs):
                                     del item2['redirect_configuration']
                                 if item2['rewrite_rule_set']:
                                     id = item2['rewrite_rule_set']
-                                    id = id if is_valid_resource_id(id) else rerite_rule_set_id(self.subscription_id,
+                                    id = id if is_valid_resource_id(id) else rewrite_rule_set_id(self.subscription_id,
                                                                                                 kwargs['resource_group'],
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]: lines 19334 1935 and 1936 continuation line under-indented for visual indent

File: plugins/modules/azure_rm_appgateway.py [hunk 10]
```diff
@@ -1614,7 +1948,7 @@ def exec_module(self, **kwargs):
                             del item['default_redirect_configuration']
                         if item['default_rewrite_rule_set']:
                             id = item['default_rewrite_rule_set']
-                            id = id if is_valid_resource_id(id) else rerite_rule_set_id(self.subscription_id,
+                            id = id if is_valid_resource_id(id) else rewrite_rule_set_id(self.subscription_id,
                                                                                         kwargs['resource_group'],
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]: lines 1952, 1953 and 1954 continuation line under-indented for visual indent

File: plugins/modules/azure_rm_appgateway.py [hunk 11]
```diff
@@ -1665,7 +1999,7 @@ def exec_module(self, **kwargs):
                             item['url_path_map'] = {'id': id}
                         if item.get('rewrite_rule_set'):
                             id = item.get('rewrite_rule_set')
-                            id = id if is_valid_resource_id(id) else rerite_rule_set_id(self.subscription_id,
+                            id = id if is_valid_resource_id(id) else rewrite_rule_set_id(self.subscription_id,
                                                                                         kwargs['resource_group'],
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]: lines 2003, 2004 and 2005 continuation line under-indented for visual indent

File: plugins/modules/azure_rm_appgateway.py [hunk 12]
```diff
@@ -2022,10 +2364,40 @@ def rerite_rule_set_id(subscription_id, resource_group_name, appgateway_name, na
         name
     )
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py:2369]: expected 2 blank lines, found 1
```suggestion


```

File: plugins/modules/azure_rm_appgateway.py [hunk 13]
```diff
@@ -612,6 +652,102 @@
...
+                description:
+                    - Maximum request body size in Kb for WAF.
+                type: int
+            request_body_check:
+                description:
+                    - Whether allow WAF to check request Body.
+                type: bool
+            rule_set_type:
+                description:
+                    - The type of the web application firewall rule set. Possible values are: 'OWASP'.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py]:
```suggestion
                    - The type of the web application firewall rule set.
                    - Possible values are 'OWASP'.
```

File: plugins/modules/azure_rm_appgateway.py [hunk 14]
```diff
@@ -612,6 +652,102 @@
...
+                type: int
+            min_capacity:
+                description:
+                    - Lower bound on number of Application Gateway capacity.
+                type: int
+    enable_http2:
+        version_added: "1.14.0"
+        description:
+            - Whether HTTP2 is enabled on the application gateway resource.
+        type: bool
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_appgateway.py:673]:
```suggestion
        type: bool
        default: False
```

**Review by Fred-sun — COMMENTED** [review]
Small change request!

**Review by xuzhang3 — APPROVED** [review]

**Comment by xuzhang3** [comment]
@petitalotc  LGTM

**Pull request merged**